edge-hub/watchdog/linux/src/main.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    io::Error,
+    process::{exit, Child, Command, Stdio},
+    sync::atomic::{AtomicBool, Ordering},
+    sync::Arc,
+    thread,
+    time::Duration,
+};
+
+use nix::sys::signal::{self, Signal};
+use nix::unistd::Pid;
+use signal_hook::{flag, SIGINT, SIGTERM};
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+const PROCESS_SHUTDOWN_TOLERANCE_SECS: u64 = 5;
+const PROCESS_POLL_INTERVAL_SECS: u64 = 1;
+
+struct ChildProcess {
+    pub name: String,
+    pub process: Child,
+}
+
+// TODO: find documentation standards for func param
+impl ChildProcess {
+    pub fn run_child_process(
+        name: String,
+        command: &mut Command,
+        has_parent_started_shutdown: Arc<AtomicBool>,
+    ) -> Arc<AtomicBool> {
+        // start child process before spawning new thread because it will get freed after the caller runs this function
+        // this makes lifetimes easier to deal with
+        let child = match command.spawn() {
+            Ok(child) => {
+                info!("Launched {:?} process with pid {:?}", name, child.id());
+                child
+            }
+            Err(e) => {
+                error!("Failed to start {:?} process. {:?}", name, e);
+                exit(2);
+            }
+        };
```

Comment:
Is having this child process created before the thread spawns a good idea?

Reply:
I don't see a problem with that.

Reply:
@dmolokanov This is what I tried to do initially, but since the command that I pass into this ```run_child_process function``` is freed at the end of the call to this same function, I had difficulty giving it to another thread. I thought that it was better to avoid the lifetimes and do what was needed with the command immediately.